optuna/distributions.py

```diff
@@ -79,16 +78,37 @@ def _contains(self, param_value_in_internal_repr):
 
         raise NotImplementedError
 
-    @abc.abstractmethod
     def _asdict(self):
         # type: () -> Dict
 
-        raise NotImplementedError
+        return self.__dict__
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+
+        if not isinstance(other, type(self)):
+            return False
+
+        return self.__dict__ == other.__dict__
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        # type: () -> int
+
+        return hash(tuple([self.__class__] + sorted(self.__dict__.items())))
```

Comment:
The mypy check failed at this line (see https://circleci.com/gh/pfnet/optuna/21968?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link).
IMO, `self.__class__` is not necessary because the current `master` does not check class information, and it only takes the attributes into accounts.

```python
>>> hash(optuna.distributions.UniformDistribution(1, 2))
3713081631934410656
>>> hash(optuna.distributions.LogUniformDistribution(1, 2))
3713081631934410656
>>> hash(optuna.distributions.LogUniformDistribution(1, 2)) == hash(optuna.distributions.LogUniformDistribution(1, 2))
True
```

Reply:
Thanks. I adopted your suggestion (https://github.com/pfnet/optuna/pull/573/commits/7fb958518d731090ba3c80634b1778b43578aaf4).